### PR TITLE
Feature/nakijin 489 add accommodations merge endpoint

### PR DESCRIPTION
**Pull request by oleg-n-k**
Add an endpoint to merge accommodation groups from AccommodationUncertainMatches relation

### CODE REVIEW

**Review by AlexeyShok — APPROVED** [review]

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+
+            var (_, isGetAccommodationsFailure, accommodationsData, getAccommodationsError) =
+                await GetAccommodations(accommodationIdGroups.SelectMany(a => a).ToList(), cancellationToken);
+
+            if (isGetAccommodationsFailure)
+                return Result.Failure(getAccommodationsError);
+
+            return await MergeAccommodationGroups(accommodationIdGroups, accommodationsData)
+                .BindWithTransaction(_dbContext, () => DeactivateUncertainMatches(uncertainMatchesRelations.Select(r => r.Id).ToList(), cancellationToken))
+                .Tap(() => ProcessAccommodationMappingGroups(accommodationIdGroups, cancellationToken))
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:60]: Accommodation mappings are processed in Merge, why needed new functionality ?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+                .Tap(() =>
+                {
+                    // Save all data to db
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:63]: why not saved data for now ?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+                foreach (var accommodationMapping in newAccommodationMappings)
+                {
+                    _mappingsCache.AddOrReset(accommodationMapping.HtId, accommodationMapping.MappedHtIds);
+                }
+            }
+        }
+
+
+        private async Task PublishChanges(List<RichAccommodationDetails> accommodations)
+        {
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:461]: I think needed publish only removed accommodations

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+            await _accommodationChangePublisher.PublishRemoved(accommodationIdToMerge);
+
+            await _accommodationChangeLogCollectorService.SaveCollectedEventLogs(cancellationToken);
+            return Result.Success();
+        }
+
+
+        private Result<List<List<int>>> RetrieveAccommodationIdGroups(List<AccommodationUncertainMatchesHtIds> accommodationUncertainMatchesHtIds)
+        {
+            var retrieveAccommodationIdResults = accommodationUncertainMatchesHtIds.Select(accommodationUncertainMatchesHtIdGroup
```
[the remark this answers is not in the document]
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:147]: fixed

File: HappyTravel.Nakijin.Api/Controllers/AdministratorControllers/AccommodationMergeController.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    public AccommodationMergeController(IAccommodationMergeService accommodationMergeService)
+    {
+        _accommodationMergeService = accommodationMergeService;
+    }
+
+
+    /// <summary>
+    /// Merges accommodations from an uncertain matches relation
+    /// </summary>
+    /// <param name="accommodationToMergeRequest"></param>
```
[the remark this answers is not in the document]
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Controllers/AdministratorControllers/AccommodationMergeController.cs]: fixed

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+        private Result MergeAccommodations(List<int> accommodationIds, Dictionary<int, RichAccommodationDetails> accommodationsData)
+        {
+            var sourceAccommodationId = accommodationIds.First();
+            var sourceAccommodations = accommodationsData[sourceAccommodationId];
+
+            foreach (var accommodationIdToMerge in accommodationIds.Skip(1))
+            {
+                var accommodationToMerge = accommodationsData[accommodationIdToMerge];
+
+                var (_, isMergeFailure, mergeError) = Merge(sourceAccommodations, accommodationToMerge);
```
[the remark this answers is not in the document]
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs]: discussed, fixed

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs (hunk 1 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:63]: discussed

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs (hunk 2 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs]: discussed

Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs]: commentary has been added

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs (hunk 3 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:484]: discussed, fixed

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+                .Where(r => r.IsFailure)
+                .ToList();
+
+            return retrieveAccommodationIdFailureResults.Any()
+                ? Result.Failure<List<List<int>>>(string.Join(Environment.NewLine, retrieveAccommodationIdFailureResults.Select(i => i.Error)))
+                : retrieveAccommodationIdResults.Select(r => r.Value).ToList();
+        }
+
+
+        private Result CheckIfRequestValid(List<List<int>> accommodationIdGroups, List<SlimAccommodationUncertainMatches> slimAccommodationUncertainMatches)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationMergeService.cs:179]: I think in method name must be indicated what request model validated

**Pull request merged**